### PR TITLE
feat(hip,aiter): add AITER backend to rope cos/sin-cache ops

### PR DESCRIPTION
**Pull request by demandal25**
## Summary

Adds an AITER backend to the cos/sin-cache RoPE entry points on ROCm (gfx942/gfx950), mirroring the existing `rmsnorm` / `fused_add_rmsnorm` / `silu_and_mul` AITER backends. `backend="auto"` (default) is shape-aware: it picks AITER only where it is both faster and precise enough, and stays on the native JIT kernel otherwise.

#### `flashinfer/rope.py`

- **`apply_rope_with_cos_sin_cache`** and **`apply_rope_with_cos_sin_cache_inplace`** — gain a `backend: str = "auto"` parameter. On HIP, `backend="aiter"` routes to AITER's `rope_cached_positions_2c` kernel; `"native"` keeps the existing JIT path; `"auto"` selects per the policy below. Unknown backends raise `ValueError`.
- **`_auto_select_rope_backend(query, key)`** (new) — shape/dtype/device-aware selector (see policy table).
- **`_apply_rope_cos_sin_cache_aiter`** (new helper) — adapts FlashInfer's tensor formats to AITER's: splits the `(max_seq_len, rotary_dim)` float32 `cos||sin` cache into two separate `(max_seq_len, 1, 1, rotary_dim//2)` tables in the query dtype with `reuse_freqs_front_part=True`; reshapes Q/K to AITER's `(1, nnz, num_heads, head_dim)` layout and rotates only the leading `rotary_dim` slice (`nope_first=False`), so partial-rotary (`rotary_dim < head_size`) works; and calls AITER's lower-level `rope_cached_positions_2c_fwd_impl` with **distinct in/out tensors** so the out-of-place path writes straight into the fresh output with **no Q/K copy** (AITER's `_inplace` wrapper is just `_impl(x,y,x,y,...)`). For partial rotary the untouched nope tail is copied across only when the output is a fresh tensor.
- **`_aiter_rope_cos_sin`** (new helper) — memoizes the dtype-converted cos/sin tables, keyed by the cache tensor's `id()` with a `weakref.finalize` evictor (and a `weakref.ref` identity check on hit to guard against `id()` reuse) so they never outlive the source and the full-table conversion runs once.

#### `README.md`

- Feature matrix RoPE row + AITER Support section document the auto policy.

#### `tests/rocm_tests/test_rope_aiter_hip.py` (new)

- AITER-vs-native-reference across dtypes / shapes / NEOX styles (incl. partial rotary), inplace == non-inplace equivalence, the auto selector (incl. mixed-dtype fallback), unknown-backend raises, mixed-dtype / odd-rotary / oversized-rotary raise under explicit `aiter`, and non-contiguous positions are normalized (no process abort).

## Benchmark results

Measured on gfx942 (MI300X), `q32/k8`, `hd128`, `rot128`, CUDA-event timed, warm JIT + warm cos/sin memo. Time is mean µs per call; ratio is native/aiter (>1 = AITER faster).

**Out-of-place** (`apply_rope_with_cos_sin_cache`) — zero-copy via the `_impl` entry point makes AITER win across the board at large nnz (bf16 shown; the copy in the original `_inplace`-based approach had made these <1x):

| nnz | native | aiter | native/aiter |
| ---: | ---: | ---: | ---: |
| 8 (decode) | 17.0 | 13.1 | 1.29x |
| 2048 | 34.1 | 13.0 | 2.63x |
| 8192 | 93.8 | 39.6 | 2.37x |
| 32768 | 488.4 | 176.3 | 2.77x |

**Inplace** (`apply_rope_with_cos_sin_cache_inplace`, bf16) — crosses over around 1024–1536 tokens:

| nnz | native | aiter | native/aiter | winner |
| ---: | ---: | ---: | ---: | :--- |
| 8 (decode) | 7.2 | 10.8 | 0.67x | native |
| 1024 | 12.8 | 13.0 | 0.98x | ~tie |
| 2048 | 24.6 | 19.6 | 1.26x | AITER |
| 32768 | 364.6 | 220.9 | 1.65x | AITER |

`max_seq` (4K→128K) has no measurable effect at fixed nnz — the cos/sin memoization removes the per-call table conversion, so cache size is irrelevant.

## Architecture / design notes

**`backend="auto"` policy** (`_auto_select_rope_backend`):

| Condition | Resolves to | Why |
| :--- | :--- | :--- |
| dtype != fp16 | `native` | bf16 AITER error ~5e-2 vs native ~3e-2 (tolerance edge); fp16 ~7e-3 is safe |
| query/key dtype mismatch | `native` | AITER rotates both with one cos/sin table; native handles mixed dtypes — auto must not raise |
| nnz < 2048 | `native` | below crossover, native's lower launch overhead wins |
| aiter unsupported/unimportable | `native` | best-effort probe; auto never raises |
| else (fp16 + nnz≥2048 + supported) | `aiter` | measured ~1.3–2.8x, both inplace and out-of-place |

**Why the `_impl` entry point.** AITER's public `_inplace`/`_fwd` wrappers force either out==in or an internal allocation. `_impl` (which both wrappers delegate to) exposes distinct in/out tensors, giving zero-copy out-of-place without reaching below AITER's stable kernel boundary.

**Why only the cos/sin-cache functions.** AITER's `rope_cached_*` API takes a precomputed cos/sin cache, matching `apply_rope_with_cos_sin_cache` 1:1. The other rope entry points compute frequencies on the fly and have no corresponding AITER cached kernel.

**Guards on the explicit `backend="aiter"` path** (the native path tolerates these; AITER does not):

| Guard | Reason |
| :--- | :--- |
| device must be gfx942/gfx950 | otherwise a cryptic import/JIT failure |
| `aiter` import wrapped | clear `ValueError` if the package is missing |
| query/key must share a dtype | AITER rotates both with one cos/sin table built in the query dtype |
| `rotary_dim` even and ≤ head_size | the cos||sin split and the rotated slice would otherwise be malformed |
| positions → contiguous int64 | a strided positions tensor trips a C `assert` that aborts the process |

## Test plan

- [x] `pytest tests/rocm_tests/test_rope_aiter_hip.py` (passed on gfx942)
- [x] `pytest tests/rocm_tests/test_rope_hip.py` (native path unbroken; native ↔ aiter agree within tolerance)
- [x] `pytest tests/rocm_tests/test_rope_hip.py tests/rocm_tests/test_rope_aiter_hip.py -n auto --reruns 2 -m "not slow"` (0 failures)
- [x] `pre-commit run -a`

### CODE REVIEW

File: flashinfer/rope.py [hunk 1]
```diff
@@ -1181,6 +1297,25 @@ def apply_rope_with_cos_sin_cache(
...
+                cos_sin_cache=cos_sin_cache,
+                positions=positions,
+                head_size=head_size,
+                is_neox=is_neox,
+            )
+            return query_out, key_out
+        if _backend != "native":
+            raise ValueError(
+                f"Unknown backend {backend!r}; expected one of 'auto', 'native', 'aiter'."
+            )
```
[the remark this answers is not in the document]
Comment (demandal25) [reply flashinfer/rope.py:1392]: Not addressing this: `amd-flashinfer` is an AMD-ROCm-only package — `IS_HIP` is always `True` on every supported target (gfx942/gfx950), and it never ships on CUDA. The unknown-backend `ValueError` and the `backend="aiter"` handling already live inside the `if IS_HIP:` block, so on real hardware misconfigurations *are* caught. Unconditional validation would only change behavior on a non-HIP build we don't produce. Note the cited reference is mixed: `activation.py` validates unconditionally, but `norm.py` validates inside `if IS_HIP:` exactly like this code.

File: flashinfer/rope.py [hunk 2]
```diff
@@ -1227,13 +1363,41 @@ def apply_rope_with_cos_sin_cache_inplace(
...
+                cos_sin_cache=cos_sin_cache,
+                positions=positions,
+                head_size=head_size,
+                is_neox=is_neox,
+            )
+            return
+        if _backend != "native":
+            raise ValueError(
+                f"Unknown backend {backend!r}; expected one of 'auto', 'native', 'aiter'."
+            )
```
[the remark this answers is not in the document]
Comment (demandal25) [reply flashinfer/rope.py:1480]: Not addressing this, same reason as the non-inplace variant above: this is an AMD-ROCm-only package where `IS_HIP` is always `True` on supported hardware, so the in-`IS_HIP` validation already catches misconfiguration on every real target. The non-HIP fallback path this comment describes is unreachable in shipped builds.

File: flashinfer/rope.py [hunk 3]
```diff
@@ -28,6 +30,111 @@ def get_rope_module():
...
+            # query dtype; the native path tolerates mixed dtypes by rotating
+            # in float32, but AITER cannot.
+            raise ValueError(
+                "AITER rope backend requires query and key to share a dtype; "
+                f"got query={query.dtype}, key={key.dtype}. Use backend='native'."
+            )
+
+        nnz = query.shape[0]
+        rotary_dim = cos_sin_cache.shape[-1]
+        cos, sin = _aiter_rope_cos_sin(cos_sin_cache, query.dtype)
```
[the remark this answers is not in the document]
Comment (demandal25) [reply flashinfer/rope.py]: Fixed in 2897bb74's follow-up. Added explicit validation in `_apply_rope_cos_sin_cache_aiter` before the split: `rotary_dim` (= `cos_sin_cache.shape[-1]`) must be even (it stacks cos||sin) and must not exceed `head_size`, each raising a clear `ValueError`. Covered by `test_rope_aiter_odd_rotary_dim_raises` and `test_rope_aiter_rotary_dim_exceeds_head_size_raises`.

File: flashinfer/rope.py [hunk 4]
```diff
@@ -28,6 +30,111 @@ def get_rope_module():
...
+        _aiter_rope_ops().rope_cached_positions_2c_fwd_inplace(
+            q_view[..., :rotary_dim],
+            k_view[..., :rotary_dim],
+            cos,
+            sin,
+            pos,
+            0 if is_neox else 1,  # rotate_style: 0=NEOX, 1=GPT-J
+            True,  # reuse_freqs_front_part (cos/sin are rotary_dim//2 sized)
+            False,  # nope_first
+        )
```
[the remark this answers is not in the document]
Comment (demandal25) [reply flashinfer/rope.py]: Fixed. The AITER-ops resolution in `_apply_rope_cos_sin_cache_aiter` is now wrapped: a failed `_aiter_rope_ops()` import is caught and re-raised as a clear `ValueError("backend='aiter' requires the aiter package, which failed to import: ...")`, matching the `activation.py` convention. (The `backend="auto"` selector already probes import availability and falls back to native, so this only affects the explicit `backend="aiter"` opt-in.)

File: flashinfer/rope.py [hunk 5]
```diff
@@ -28,6 +30,152 @@ def get_rope_module():
...
+            return "native"
+        if query.dtype != torch.float16:
+            return "native"
+        if query.shape[0] < _AITER_ROPE_MIN_TOKENS:
+            return "native"
+        from .aiter_utils import is_aiter_supported
+
+        if not is_aiter_supported(query.device):
+            return "native"
+        return "aiter"
```
[the remark this answers is not in the document]
Comment (demandal25) [reply flashinfer/rope.py:78]: Already addressed in bb17a021 (this review was against cf8e5d5c). `_auto_select_rope_backend` now ends with a best-effort probe: after the arch check it calls `_aiter_rope_ops()` inside a try/except and returns "native" on any import/load failure, so `backend="auto"` falls back gracefully — matching `activation._auto_select_silu_and_mul_backend`. Covered by the new `test_rope_auto_falls_back_when_aiter_unimportable`.

File: flashinfer/rope.py [hunk 6]
```diff
@@ -28,6 +30,152 @@ def get_rope_module():
...
+        cached = _aiter_cos_sin_tables.get(key)
+        if cached is not None and cached[0] == dtype:
+            return cached[1], cached[2]
+        half = cos_sin_cache.shape[-1] // 2
+        cos = cos_sin_cache[:, :half].unsqueeze(1).unsqueeze(1).to(dtype)
+        sin = cos_sin_cache[:, half:].unsqueeze(1).unsqueeze(1).to(dtype)
+        if cached is None:
+            weakref.finalize(cos_sin_cache, _aiter_cos_sin_tables.pop, key, None)
+        _aiter_cos_sin_tables[key] = (dtype, cos, sin)
+        return cos, sin
```
[the remark this answers is not in the document]
Comment (demandal25) [reply flashinfer/rope.py:105]: Not addressing — this is incorrect for `weakref.finalize` (it differs from `weakref.ref` here). Per the CPython docs, "finalize objects are not GC'd until the object they are bound to is GC'd or atexit is called": `finalize` registers itself in an internal registry (`weakref.finalize._registry`) and stays alive on its own until the referent is collected. So discarding the return value does NOT disable eviction. Verified empirically: with the return value discarded (exactly as written), deleting the source tensor + gc.collect() still fires the callback and pops the entry, and while the tensor is alive the entry is retained. The cache therefore does not grow unbounded. (`id()` reuse is separately guarded by the `weakref.ref` identity check on hit, added in bb17a021.)

File: tests/rocm_tests/test_rope_aiter_hip.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    # Out-of-place always native (the Q/K copy erases AITER's throughput win).
+    assert _auto_select_rope_backend(q_fp16_big, inplace=False) == "native"
+
+    # bf16 always native (precision: ~5e-2 vs native ~3e-2).
+    q_bf16_big = torch.randn(big, 128, dtype=torch.bfloat16, device=device)
+    assert _auto_select_rope_backend(q_bf16_big, inplace=True) == "native"
+
+    # Below the token threshold, native's lower launch overhead wins.
+    q_fp16_small = torch.randn(small, 128, dtype=torch.float16, device=device)
+    assert _auto_select_rope_backend(q_fp16_small, inplace=True) == "native"
```
[the remark this answers is not in the document]
Comment (demandal25) [reply tests/rocm_tests/test_rope_aiter_hip.py]: Fixed. Added `test_rope_auto_falls_back_when_aiter_unimportable` in test_rope_aiter_hip.py: it monkeypatches `flashinfer.rope._aiter_rope_ops` to raise ImportError and asserts `_auto_select_rope_backend(q, k)` returns "native" (and that it returns "aiter" before the patch, as a sanity check). Note: the selector signature is `(query, key)` — the `inplace` parameter was removed in bb17a021 when the zero-copy `_impl` path made out-of-place benefit too, so auto no longer distinguishes inplace vs out-of-place.

File: tests/rocm_tests/test_rope_aiter_hip.py [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+import torch
+
+import flashinfer
+from flashinfer.aiter_utils import is_aiter_supported
+from tests.test_helpers.rope_reference import RotaryEmbedding
+
+pytestmark = pytest.mark.skipif(
+    not is_aiter_supported(torch.device("cuda:0")),
+    reason="AITER backend requires gfx942/gfx950",
+)
```
[the remark this answers is not in the document]
Comment (demandal25) [reply tests/rocm_tests/test_rope_aiter_hip.py]: Fixed in 390a1ae3. Switched the module `pytestmark` from the arch-only `is_aiter_supported` check to `tests.test_helpers.test_helpers.requires_aiter`, which skips when the arch is unsupported OR the `aiter` package is not importable — matching the other `test_*_aiter_hip.py` modules. On a supported GPU without aiter installed these tests now skip instead of failing.
